File: client/csharp/example/connect/test/cli/AsyncMenu.cs
using System;
using System.Collections.Generic;
using Connect;

namespace example
{
  class AsyncMenu {
    private Menu menu;
    private static ConnectSvc connectSvc;

    public AsyncMenu(ConnectSvc svc) {
      connectSvc = svc;

      MenuItem[] items = new MenuItem[4];
      items[0] = new MenuItem("1", "Add async connections", new MenuCallback(AddAsyncConnection), false);
      items[1] = new MenuItem("2", "Delete async connections", new MenuCallback(DeleteAsyncConnection), false);
      items[2] = new MenuItem("3", "Refresh the connection list", new MenuCallback(ShowAsyncConnection), false);
      items[3] = new MenuItem("q", "Return to Main Menu", null, true);      

      menu = new Menu(items);
    }

    public void Show() {
      ShowAsyncConnection();

      menu.Show("Async Menu");
    }

    public static void ShowAsyncConnection() {
      Console.WriteLine("Getting the async connections...");

      try {
        var devList = connectSvc.GetDeviceList();
        var asyncConns = new List<DeviceInfo>();

        for(int i = 0; i < devList.Count; i++) {
          if(devList[i].AutoReconnect) {
            asyncConns.Add(devList[i]);
          }
        }

        Console.WriteLine();
        Console.WriteLine("***** Async Connections: {0}", asyncConns.Count);
        for(int i = 0; i < asyncConns.Count; i++) {
          Console.WriteLine(asyncConns[i]);
        }

        Console.WriteLine();
      } catch(Exception e) {
        Console.WriteLine("Cannot get the async connections: {0}", e);
      }
    }

    public static void AddAsyncConnection() {
      List<AsyncConnectInfo> asyncConns = new List<AsyncConnectInfo>();

      while(true) {
        Console.Write(">> Enter the device ID (Press just ENTER if no more device): ");
        string input = Console.ReadLine();

        if(input.Trim().Equals("")) {
          break;
        }

        uint devID = 0;

        if(!UInt32.TryParse(input, out devID)) {
          Console.WriteLine("Invalid device ID: {0}", input);
          return;
        }

        var connInfo = MainMenu.GetConnectInfo();

        if(connInfo == null) {
          return;
        }

        AsyncConnectInfo asyncInfo = new AsyncConnectInfo{ DeviceID = devID, IPAddr = connInfo.IPAddr, Port = connInfo.Port, UseSSL = connInfo.UseSSL };
        asyncConns.Add(asyncInfo);
      }

      if(asyncConns.Count == 0) {
        Console.WriteLine("You have to enter at least one async connection");
        return;
      }

      try {
        Console.WriteLine("Adding async connections...");
        connectSvc.AddAsyncConnection(asyncConns.ToArray());
        ShowAsyncConnection();
      } catch(Exception e) {
        Console.WriteLine("Cannot add the async connections: {0}", e);
      }
    }

    public static void DeleteAsyncConnection() {
      uint[] deviceIDs = Menu.GetDeviceIDs();

      if(deviceIDs == null) {
        return;
      }

      try {
        Console.WriteLine("Deleting async connections...");
        connectSvc.DeleteAsyncConnection(deviceIDs);
        ShowAsyncConnection();
      } catch(Exception e) {
        Console.WriteLine("Cannot delete the async connections: {0}", e);
      }
    }
  }
}

File: client/python/example/connect/test/cli/asyncMenu.py
import grpc

import connect_pb2

from example.cli.menu import MenuItem, Menu
from example.cli.input import getUserInput, UserInput, getDeviceIDs

class AsyncMenu:
  menu = None
  connectSvc = None

  def __init__(self, connectSvc):
    self.connectSvc = connectSvc

    menuItems = [
      MenuItem('1', 'Add async connections', self.addAsyncConnection, False),
      MenuItem('2', 'Delete async connections', self.deleteAsyncConnection, False),
      MenuItem('3', 'Refresh the connection list', self.showAsyncConnection, False),
      MenuItem('q', 'Return to Main Menu', None, True),
    ]

    self.menu = Menu('Async Menu', menuItems)

  def showAsyncConnection(self):
    connInfos = []

    try:
      devList = self.connectSvc.getDeviceList()

      for dev in devList:
        if dev.autoReconnect:
          connInfos.append(dev)

      print(f'\n***** Async connections: {len(connInfos)}', flush=True)
      print(connInfos, flush=True)

      return connInfos
    
    except grpc.RpcError as e:
      print(f'Cannot show the async connections: {e}')      
      return []

  def getAsyncConnectionInfo(self):
    connInfos = []

    userInputs = [
      UserInput('Enter the IP address of the device', ''),
      UserInput('Enter the port of the device (default: 51211)', '51211'),
      UserInput('Use SSL y/n (default: n)', 'n')
    ]

    while True:
      devIDStr = input('>> Enter the device ID (Press just ENTER if no more device): ')

      if devIDStr.strip() == '':
        break

      try:
        devID = int(devIDStr)
      except ValueError as verr:
        print(f'Invalid device ID {devIDStr}: {verr}')
        break

      inputs = getUserInput(userInputs)

      try:
        port = int(inputs[1])
      except ValueError as verr:
        print(f'Invalid port number {inputs[1]}: {verr}')      
        break

      useSSL = False
      if inputs[2].strip().lower() == 'y':
        useSSL = True

      connInfo = connect_pb2.AsyncConnectInfo(deviceID=devID, IPAddr=inputs[0], port=port, useSSL=useSSL)
      connInfos.append(connInfo)

    return connInfos

  def addAsyncConnection(self):
    connInfos = self.getAsyncConnectionInfo()

    if len(connInfos) == 0:
      print('No device to add', flush=True)
      return

    try:
      self.connectSvc.addAsyncConnection(connInfos)
      self.showAsyncConnection()

    except grpc.RpcError as e:
      print(f'Cannot add async connections: {e}')    

  def deleteAsyncConnection(self):
    deviceIDs = getDeviceIDs()

    if len(deviceIDs) == 0:
      print('No device to delete', flush=True)
      return

    try:
      self.connectSvc.deleteAsyncConnection(deviceIDs)
      self.showAsyncConnection()

    except grpc.RpcError as e:
      print(f'Cannot delete async connections: {e}')    
  
  def show(self):
    self.showAsyncConnection()
    self.menu.show()

